### PR TITLE
chore: unindexed fragment can handle mulitple deltas

### PR DESCRIPTION
**Pull request by eddyxu**
BREAKING CHANGE: 
* change the structure of `index_statistics` return to include a list of index deltas.
* Do not return the cache hit rate because it is a global counter, not per-index status.
* Support f16/f32/f64 as IVF centroids.

### CODE REVIEW

File: rust/lance/src/index.rs [hunk 1]
```diff
@@ -346,57 +348,65 @@ impl DatasetIndexExt for Dataset {
...
+            .iter()
+            .map(|idx| idx.statistics())
+            .collect::<Result<Vec<_>>>()?;
 
-        let (count_rows, count_unindexed_rows) =
-            futures::try_join!(count_rows, count_unindexed_rows)?;
+        let unindexed_fragments = self.unindexed_fragments(index_name).await?;
+        let num_unindexed_rows = unindexed_fragments
+            .iter()
+            .map(|f| f.num_rows().unwrap_or_default())
```
Comment (westonpace) [inline rust/lance/src/index.rs:384]: `unwrap_or_default` will return `0` for any `None`.  We should probably error out instead if it is `None`.

File: rust/lance-index/src/traits.rs [hunk 2]
```diff
@@ -64,10 +64,21 @@ pub trait DatasetIndexExt {
...
-    async fn load_index_by_name(&self, name: &str) -> Result<Option<Index>> {
-        self.load_indices()
-            .await
-            .map(|indices| indices.iter().find(|idx| idx.name == name).cloned())
+    ///
+    /// Returns
+    /// -------
+    /// - `Ok(indices)`: if the index exists, returns the index.
+    /// - `Ok(empty)`: if the index does not exist.
+    /// - `Err(e)`: if there is an error loading indices.
```
Comment (westonpace) [inline rust/lance-index/src/traits.rs:72]:
```suggestion
    /// - `Ok(Some(indices))`: if the index exists, returns all deltas in the index.
    /// - `Ok(None)`: if the index does not exist.
    /// - `Err(e)`: if there is an error loading indices.
```

File: rust/lance-index/src/traits.rs [hunk 3]
```diff
@@ -77,15 +88,5 @@ pub trait DatasetIndexExt {
...
-    /// Count the rows that are not indexed by the given index.
-    ///
-    /// TODO: move to [DatasetInternalExt]
-    async fn count_unindexed_rows(&self, index_name: &str) -> Result<Option<usize>>;
-
-    /// Count the rows that are indexed by the given index.
-    ///
-    /// TODO: move to [DatasetInternalExt]
-    async fn count_indexed_rows(&self, index_name: &str) -> Result<Option<usize>>;
+    async fn index_statistics(&self, index_name: &str) -> Result<serde_json::Value>;
```
Comment (westonpace) [inline rust/lance-index/src/traits.rs:91]: This puts `serde_json` on our public API.  It's just a minor pain (e.g. users have to make sure they are using the same version of `serde_json`).  Is there a particular reason you didn't want to use `String` here?  Maybe for composing statistics from child indices?

I don't think this is something we need to fix for this PR, just noting it.  Maybe we can just make a follow-up issue for it.

File: rust/lance-index/src/traits.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance-index/src/traits.rs:72]: This does not return Option<Vec<Index>>

**Pull request merged**